src/server.rs

```diff
@@ -311,7 +335,18 @@ impl Read for Encoder {
                     ref mut chunk,
                     is_last,
                 } => {
-                    bytes_read += ready!(Pin::new(chunk).poll_read(cx, &mut buf))?;
+                    let inner_poll_result = Pin::new(chunk).poll_read(cx, &mut buf);
+                    bytes_read += match inner_poll_result {
+                        Poll::Ready(Ok(n)) => n,
+                        Poll::Pending => {
+                            if bytes_read == 0 {
+                                return Poll::Pending;
+                            } else {
+                                break;
+                            }
+                        }
+                        e => return e,
```

Comment:
Even though this is more typing, this should be easier to follow since all left-hand match arms then use fully qualified types, rather than a named catch-all argument as is the case now:

```suggestion
                        Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
```